Show selected end of line in TextEdit

aXu-AP opened the pull request:
Fixes #72265
Continues selection box at the end of line if eol/carriage return is selected.
Should be working well, works even with rtl text mixed in.
![2023-01-29 23-38-28](https://user-images.githubusercontent.com/1621768/215357592-78059e1c-e679-4f28-8650-8aad1c83053a.gif)
Also tested to be working with rtl mode.

aXu-AP committed:
Show selected end of line in TextEdit

Continues selection box at the end of line if eol/carriage return is selected.

Paulb23 commented on scene/gui/text_edit.cpp:
```diff
@@ -1102,6 +1090,38 @@ void TextEdit::_notification(int p_what) {
+
+							// End of line selected
+							if (line < get_selection_to_line(c)) {
+								float char_w = font->get_char_size(' ', font_size).width;
```
We should cache this as `get_char_size` can be a slow if called a lot.

Paulb23 commented on scene/gui/text_edit.cpp:
```diff
@@ -1102,6 +1090,38 @@ void TextEdit::_notification(int p_what) {
 							}
+
+							// End of line selected
+							if (line < get_selection_to_line(c)) {
```
Would it worth pushing back into `Vector<Vector2> sel` rather then duplicating the validation code?

aXu-AP replied:
Yes it would. Way neater.

Paulb23 commented on scene/gui/text_edit.cpp:
```diff
@@ -1083,11 +1071,31 @@ void TextEdit::_notification(int p_what) {
+								if (rtl) {
+									// Find the leftmost character's bounds
+									for (int j = 0; j < char_bounds.size(); j++) {
+										last_char_offset = MIN(last_char_offset, char_bounds[j].x);
```
Won't this always be `0`?

aXu-AP replied:
No, if you align text right, which makes sense in an rtl context. I don't know if there's any way to see this in code editor, but it works if you set TextEdit layout direction as rtl.
However, I just realized that you can use right justification for ltr text too, I should bind this to the `text_direction` property, not layout direction. In that case using left justified rtl text or right justified ltr text, the end of line won't be visible since there is no margin to draw them to, but I think it's a _marginal_ use case anyway 😅

Paulb23 commented on scene/gui/text_edit.cpp:
```diff
@@ -1083,11 +1071,31 @@ void TextEdit::_notification(int p_what) {
 							Vector<Vector2> sel = TS->shaped_text_get_selection(rid, sel_from, sel_to);
+							// End of line selected
+							if (line < get_selection_to_line(c)) {
+								Vector<Vector2> char_bounds = TS->shaped_text_get_selection(rid, 0, str.length());
```
Could we use something like `shaped_text_get_size` if we're calculating the entire length, rather then looping over everything?

aXu-AP replied:
I didn't know this existed, thanks! It'll simplify the code nicely.

KoBeWi commented on scene/gui/text_edit.cpp:
```diff
@@ -1083,11 +1071,31 @@ void TextEdit::_notification(int p_what) {
 							int sel_from = (line > get_selection_from_line(c)) ? TS->shaped_text_get_range(rid).x : get_selection_from_column(c);
 							int sel_to = (line < get_selection_to_line(c)) ? TS->shaped_text_get_range(rid).y : get_selection_to_column(c);
 							Vector<Vector2> sel = TS->shaped_text_get_selection(rid, sel_from, sel_to);
+							// End of line selected
```
```suggestion
							// End of line selected.
```
Same below.

aXu-AP commented:
Updated now. I changed from checking if layout direction is rtl to checking the text direction itself (since you could use right justification for ltr text too). However, I couldn't yet figure out how you can detect rtl text if `text_direction == TEXT_DIRECTION_AUTO`.

Paulb23 commented:
I think checking `if rtl` as before should be okay cc @bruvzg?

Otherwise looks good.

bruvzg reviewed (commented):
There's no alignment support in TextEdit.

The only thing that should be checked is `layout_direction` (which is `if (rtl)`), it's the only thing that is determining the alignment.

The real end of line can be anywhere in the string, but for the selection purpose it's OK to add extra selection on the left/right side depending on alignment.

`text_direction == RTL` is a "base direction" which determine BiDi reordering priority (e.g, is it an Arabic word in the middle of English text or wise versa). It's not a final direction, neither it has any effect on alignment (e.g., it won't do anything for pure Latin text).

aXu-AP commented:
Okay, I'll change it back to `if (rtl)` then. End alignment is rarely needed anyway.

> The real end of line can be anywhere in the string, but for the selection purpose it's OK to add extra selection on the left/right side depending on alignment.

For reference, Microsoft Word also adds the end of line selection at the right side of text even if it's really at the center similiarly to this implementation.

Paulb23 commented:
Tested locally seems to leave a small gap, rtl looks good though:
![image](https://user-images.githubusercontent.com/6584330/218337269-86090310-a3af-4568-977c-964e8508396a.png)

aXu-AP commented:
Interesting. I can replicate that with certain line length and zoom level. Does this happen on every line on your setup?
I tested it on older build without `shaped_text_get_size`, it seems that it didn't occur then. (grab some older artifact here: https://github.com/aXu-AP/godot/actions?page=2)

aXu-AP commented:
I reverted back to using `shaped_text_get_selection` since `shaped_text_get_size` doesn't give perfect boundaries.
Sorry for leaving this for a month, my life just got quite busy for a moment 😅

Paulb23 commented:
I'm not a fan of having to loop over the entire line to draw the selection, might be worth looking at adding something to the `TextServer` to return exact rather then rounded `shaped_text_get_size`, or make it a `TextServer` behaviour?

aXu-AP commented:
One more try @Paulb23  😅
I looked into TextServer side of things but couldn't make head or tails of it. So I made a check if there's a subpixel gap between selections then it snaps to place. 

I should be a contender for the slowest contributor of the year, half a year for a simple fix and counting 🤣

aXu-AP commented:
Rebased

Paulb23 reviewed (approved):
little bit hacky, but seems okay to me, if there's no alternative.

aXu-AP commented:
Thanks, I really think this makes the script editor feel better, even if most won't notice the difference consciously.

Maybe somebody smarter will fix the underlying issue down the road. Should I open an issue for it so that it's documented? Something like: TextServer `shaped_text_get_size` and `shaped_text_get_selection` have subpixel differences.

akien-mga commented:
> Maybe somebody smarter will fix the underlying issue down the road. Should I open an issue for it so that it's documented? Something like: TextServer `shaped_text_get_size` and `shaped_text_get_selection` have subpixel differences.

CC @bruvzg

bruvzg commented:
> Maybe somebody smarter will fix the underlying issue down the road. Should I open an issue for it so that it's documented? Something like: TextServer shaped_text_get_size and shaped_text_get_selection have subpixel differences.

`shaped_text_get_size` is rounded up to ensure trimming will not cut part of string, `shaped_text_get_selection` is not, I'm not sure what's the best option to fix, I guess selection can be rounded up as well.

aXu-AP commented:
Wait, would it just work if we round up all selection rects? I'll do a test, just a moment.

aXu-AP commented:
@Paulb23 could you confirm that this solution works well and doesn't create visual problems with end of selection (as it might move the selection end by that one pixel in those cases)? I'm still hoping this to be part of 4.2 😊

I also improved on one existing comment 🤭

AThousandShips commented on scene/gui/text_edit.cpp:
```diff
@@ -1086,13 +1074,27 @@ void TextEdit::_notification(int p_what) {
+
 							for (int j = 0; j < sel.size(); j++) {
-								Rect2 rect = Rect2(sel[j].x + char_margin + ofs_x, ofs_y, sel[j].y - sel[j].x, row_height);
+								Rect2 rect = Rect2(sel[j].x + char_margin + ofs_x, ofs_y, ceil(sel[j].y - sel[j].x), row_height);
```
```suggestion
								Rect2 rect = Rect2(sel[j].x + char_margin + ofs_x, ofs_y, Math::ceil(sel[j].y - sel[j].x), row_height);
```
Use the built in methods

AThousandShips commented on scene/gui/text_edit.cpp:
```diff
@@ -1086,13 +1074,27 @@ void TextEdit::_notification(int p_what) {
+
+							// Show selection at the end of line.
+							if (line < get_selection_to_line(c)) {
+								Rect2 rect;
```
```suggestion
```
Unused

aXu-AP replied:
Thanks for noticing these 😅

akien-mga commented:
Thanks!
